hbase-server/src/main/java/org/apache/hadoop/hbase/mob/DefaultMobStoreCompactor.java

```diff
@@ -362,11 +508,375 @@ protected boolean performCompaction(FileDetails fd, InternalScanner scanner, Cel
     mobStore.updateCellsCountCompactedToMob(cellsCountCompactedToMob);
     mobStore.updateCellsSizeCompactedFromMob(cellsSizeCompactedFromMob);
     mobStore.updateCellsSizeCompactedToMob(cellsSizeCompactedToMob);
     progress.complete();
     return true;
   }
+
+  protected static String createKey(TableName tableName, String encodedName,
+      String columnFamilyName) {
+    return tableName.getNameAsString()+ "_" + encodedName + "_"+ columnFamilyName;
+  }
+
+  @Override
+  protected List<Path> commitWriter(StoreFileWriter writer, FileDetails fd,
+      CompactionRequestImpl request) throws IOException {
+    List<Path> newFiles = Lists.newArrayList(writer.getPath());
+    writer.appendMetadata(fd.maxSeqId, request.isAllFiles(), request.getFiles());
+    // Append MOB references
+    Set<String> refSet = mobRefSet.get();
+    writer.appendMobMetadata(refSet);
+    writer.close();
+    return newFiles;
+  }
+
+  private List<Path> getReferencedMobFiles(Collection<HStoreFile> storeFiles) {
+    Path mobDir = MobUtils.getMobFamilyPath(conf, store.getTableName(), store.getColumnFamilyName());
+    Set<String> mobSet = new HashSet<String>();
+    for (HStoreFile sf: storeFiles) {
+      byte[] value = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+      if (value != null) {
+        String s = new String(value);
+        String[] all = s.split(",");
+        Collections.addAll(mobSet, all);
+      }
+    }
+    List<Path> retList = new ArrayList<Path>();
+    for(String name: mobSet) {
+      retList.add(new Path(mobDir, name));
+    }
+    return retList;
+  }
+}
+
+class FileSelection implements Comparable<FileSelection> {
+
+  public final static String NULL_REGION = "";
+  private Path path;
+  private long earliestTs;
+  private Configuration conf;
+
+  public FileSelection(Path path, Configuration conf) throws IOException {
+    this.path = path;
+    this.conf = conf;
+    readEarliestTimestamp();
+  }
+
+  public  String getEncodedRegionName() {
+    String fileName = path.getName();
+    String[] parts = fileName.split("_");
+    if (parts.length == 2) {
+      return parts[1];
+    } else {
+      return NULL_REGION;
+    }
+  }
+
+  public Path getPath() {
+    return path;
+  }
+
+  public long getEarliestTimestamp() {
+    return earliestTs;
+  }
+
+  private void readEarliestTimestamp() throws IOException {
+    FileSystem fs = path.getFileSystem(conf);
+    HStoreFile sf = new HStoreFile(fs, path, conf, CacheConfig.DISABLED,
+      BloomType.NONE, true);
+    sf.initReader();
+    byte[] tsData = sf.getMetadataValue(HStoreFile.EARLIEST_PUT_TS);
+    if (tsData != null) {
+      this.earliestTs = Bytes.toLong(tsData);
+    }
+    sf.closeStoreFile(true);
+  }
+
+  @Override
+  public int compareTo(FileSelection o) {
+    if (this.earliestTs > o.earliestTs) {
+      return +1;
+    } else if (this.earliestTs == o.earliestTs) {
+      return 0;
+    } else {
+      return -1;
+    }
+  }
+
+}
+
+class Generations {
```

Comment:
I don't see any tests around Generation partitioning or the generational compaction. maybe I missed them?

Reply:
Created:
 https://issues.apache.org/jira/browse/HBASE-23189